File: AngularJS/src/org/angular2/entities/ivy/Angular2IvyComponent.java
// Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.angular2.entities.ivy;

import com.intellij.psi.PsiFile;
import org.angular2.entities.Angular2Component;
import org.angular2.entities.Angular2DirectiveSelector;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.List;
import java.util.Optional;

import static com.intellij.util.ObjectUtils.tryCast;

public class Angular2IvyComponent extends Angular2IvyDirective implements Angular2Component {

  public Angular2IvyComponent(@NotNull Angular2IvyEntityDef.Component entityDef) {
    super(entityDef);
  }

  @Nullable
  @Override
  public PsiFile getTemplateFile() {
    return null;
  }

  @NotNull
  @Override
  public List<PsiFile> getCssFiles() {
    return Collections.emptyList();
  }

  @NotNull
  @Override
  public List<Angular2DirectiveSelector> getNgContentSelectors() {
    // TODO Angular 9 d.ts metadata lacks this information
    // Try to fallback to metadata JSON information
    return Optional.ofNullable(getMetadataDirective())
      .map(directive -> tryCast(directive, Angular2Component.class))
      .map(Angular2Component::getNgContentSelectors)
      .orElseGet(Collections::emptyList);
  }
}

File: AngularJS/src/org/angular2/entities/ivy/Angular2IvyComponent.kt
// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.angular2.entities.ivy

import com.intellij.model.Pointer
import com.intellij.openapi.util.Key
import com.intellij.psi.PsiFile
import org.angular2.entities.Angular2Component
import org.angular2.entities.Angular2DirectiveKind
import org.angular2.entities.Angular2DirectiveSelector

class Angular2IvyComponent(entityDef: Angular2IvySymbolDef.Component) : Angular2IvyDirective(entityDef), Angular2Component {

  override val templateFile: PsiFile?
    get() = null

  override val directiveKind: Angular2DirectiveKind
    get() = Angular2DirectiveKind.REGULAR

  override val cssFiles: List<PsiFile>
    get() = emptyList()

  // Try to fallback to metadata JSON information - Angular 9.0.x case
  override val ngContentSelectors: List<Angular2DirectiveSelector>
    get() {
      val result = getNullableLazyValue(IVY_NG_CONTENT_SELECTORS) {
        (myEntityDef as Angular2IvySymbolDef.Component)
          .ngContentSelectors
          ?.map { createSelectorFromStringLiteralType(it) }
      }
      return result
             ?: (getMetadataDirective(typeScriptClass) as? Angular2Component)
               ?.ngContentSelectors
             ?: emptyList()
    }

  override fun createPointer(): Pointer<Angular2IvyComponent> {
    val entityDef = myEntityDef.createPointer()
    return Pointer {
      val newEntityDef = entityDef.dereference() as? Angular2IvySymbolDef.Component
      if (newEntityDef != null) Angular2IvyComponent(newEntityDef) else null
    }
  }

  companion object {

    private val IVY_NG_CONTENT_SELECTORS = Key<List<Angular2DirectiveSelector>?>("ng.ivy.content-selectors")
  }
}
